### DISCUSSION TITLE
Env vars from app.json not injected in Node.js

### DISCUSSION THREAD

**Discussion by nemanjam** [dokku/dokku#5718] [Q&A] [answered]
I have Mern app where Express server serves compiled React app. Although it seems like monorepo I only need to run Express server which is configured to serve React's build folder. I have two problems:

1. dependencies from `server/package.json` and `client/package.json` aren't installed by default because they arent in project's root folder, I have to do it in `app.json.scripts.dokku.predeploy`, that causes trouble when I just need to run `dokku ps:start my-app` I am unnecessarily reinstalling dependencies. Can I hint Dokku to install dependencies in nested folders?
2. I defined public env vars in `app.json.env` but they aren't injected in Node.js process, I logged them in js and saw logs in container, why?

I dont want to have two separate apps because Express server will serve CRA, it's a single process, I don't want to complicate with Nginx just for CRA.

**Folder structure:**

```bash
└── mern app
    ├── package.json
    ├── app.json
    ├── server (express app)
    │   ├── package.json
    │   ├── index.js
    │   ├── ...
    ├── client (create react app)
    │   ├── package.json
    │   ├── index.js
    │   ├── ...
```

**Here is my app.json:**

```js
{
    "name": "mern boilerplate",
    "description": "mern boilerplate mongo test",
    "scripts": {
      "dokku": {
        "predeploy": "npm install --prefix server && npm install --prefix client && npm run build --prefix server && npm run build --prefix client"
      }
    },
    "env": {
      "REACT_APP_BASE_URL": {
        "value": "https://mern-boilerplate.dokku.arm1.localhost3002.live"
      },
      "SITE_HOSTNAME": {
        "value": "mern-boilerplate.dokku.arm1.localhost3002.live"
      },
      "GOOGLE_CALLBACK_URL": {
        "value": "/auth/google/callback"
      },      
      "FACEBOOK_CALLBACK_URL": {
        "value": "/auth/facebook/callback"
      },
      "CLIENT_URL_PROD": {
        "value": "https://mern-boilerplate.dokku.arm1.localhost3002.live"
      },
      "SERVER_URL_PROD": {
        "value": "https://mern-boilerplate.dokku.arm1.localhost3002.live"
      },
      "MONGO_URI_PROD": {
        "generator": "echo $MONGO_URL"
      }
    },
    "addons": [
      "dokku-mongo"
    ]
  }
```

Empty env vars in Node.js in container:

```bash
Loaded env vars:  
client: --------------
REACT_APP_BASE_URL: null
CLIENT_URL_PROD: null
server: --------------
PORT: 5000
NODE_ENV: production
MONGO_URI_PROD: null
SERVER_URL_PROD: null
SITE_HOSTNAME: null
```

**Procfile:**

```bash
web: cd ./server && npm run dokku:start:server
```

**Here is the repository, `dokku-deployment` is relevant branch:**

https://github.com/nemanjam/mern-boilerplate/tree/dokku-deployment

I followed this example:

https://github.com/dokku/dokku/tree/master/tests/apps/nodejs-express

**Comment by nemanjam** [comment]
When I add `printenv &&` in `app.json` no env vars are printed during `predeploy` hook. Also later no env vars from `app.json` are defined in container. Whats the use of these vars then?

```js
{
    "name": "mern boilerplate",
    "description": "mern boilerplate mongo test",
    "scripts": {
      "dokku": {
        "predeploy": "printenv && ..."
      }
    },

```

**Comment by josegonzalez** [comment, marked as answer]
The `env` key is not supported at this time in the `app.json` file by Dokku.

**Reply by nemanjam** [reply]
So what is the solution? To set them all with `dokku config:set my-app ...`? It's not very elegant, these are all public configuration that should be set in a file and committed into git.

**Reply by josegonzalez** [reply]
So why not actually set those as variables in your code instead?

**Reply by josegonzalez** [reply]
There is an open ticket to add support for the `env` key. If thats something you're interested in [sponsoring](https://github.com/dokku/.github/blob/master/SPONSORING.md), please reach out.

**Reply by nemanjam** [reply]
> dependencies from server/package.json and client/package.json aren't installed by default because they arent in project's root folder, I have to do it in app.json.scripts.dokku.predeploy, that causes trouble when I just need to run dokku ps:start my-app I am unnecessarily reinstalling dependencies. Can I hint Dokku to install dependencies in nested folders?

Is there a way to install dependecies from nested `package.json` once on build time and not on predeploy hook?

**Reply by josegonzalez** [reply]
Assuming you are using the default nodejs buildpack, I believe it will only use the top-level package.json for it's builds. I think you have two options:

- combine the package.json files into one top-level one. it's the same app, seems a bit weird that you are managing node dependencies separately (but then trying to install both). Seems like a way to have the wrong versions of dependencies installed if the two apps have slightly different dependency trees
- You can have a top-level package.json with a custom `scripts.build` as shown [here](https://devcenter.heroku.com/articles/nodejs-support#customizing-the-build-process) that runs whatever commands you want.

**Reply by nemanjam** [reply]
Second option sounds good.
